internal/pkg/earpointer/taint.go

```diff
@@ -0,0 +1,191 @@
+//     itself is a taint source. Other fields will not be tainted.
+func (ht *heapTraversal) getStructFieldRefs(rep Reference, tp types.Type, result ReferenceSet, isTaintField func(named *types.Named, index int) bool) {
+	state := ht.partitions
+	switch tp := tp.(type) {
+	case *types.Named:
+		// Consider object of type "struct {x: *T, y *T}" where x is a
+		// taint field, and this object's heap is "{t0,t1}: [x->t2, y->t3],
+		// {t2,t4} --> t5, {t3} --> t6", then {t0,t1,t2,t4,t5} are taint sources.
+		if tt, ok := tp.Underlying().(*types.Struct); ok {
+			result[rep] = true // the current struct object is tainted
+			// Look for the taint fields.
+			for i := 0; i < tt.NumFields(); i++ {
+				f := tt.Field(i)
+				if isTaintField(tp, i) {
+					for fd, fref := range state.PartitionFieldMap(rep) {
+						if fd.Name == f.Name() {
+							result[fref] = true
+							// Mark all the subfields to be tainted.
+							ht.getFieldReferences(fref, result)
+						}
+					}
+				}
+				// Skip the non-taint fields.
+			}
+		} else {
+			ht.getStructFieldRefs(rep, tp.Underlying(), result, isTaintField)
+		}
+	case *types.Pointer:
+		if r := state.PartitionFieldMap(rep)[directPointToField]; r != nil {
+			ht.getStructFieldRefs(r, tp.Elem(), result, isTaintField)
+		} else {
+			ht.getStructFieldRefs(rep, tp.Elem(), result, isTaintField)
+		}
+	case *types.Array:
+		result[rep] = true
+		for _, r := range state.PartitionFieldMap(rep) {
+			ht.getStructFieldRefs(r, tp.Elem(), result, isTaintField)
+		}
+	case *types.Slice:
+		result[rep] = true
+		for _, r := range state.PartitionFieldMap(rep) {
+			ht.getStructFieldRefs(r, tp.Elem(), result, isTaintField)
+		}
+	case *types.Chan:
+		result[rep] = true
+		for _, r := range state.PartitionFieldMap(rep) {
+			ht.getStructFieldRefs(r, tp.Elem(), result, isTaintField)
+		}
+	case *types.Map:
+		result[rep] = true
+		for _, r := range state.PartitionFieldMap(rep) {
+			ht.getStructFieldRefs(r, tp.Elem(), result, isTaintField)
+		}
+	case *types.Basic, *types.Tuple, *types.Interface, *types.Signature:
+		// These types do not currently represent possible source types
+	}
+}
+
+// Obtains all the field references and their aliases for "ref".
+// For example, return {t0,t5,t1,t3,t4} for "{t0,t5}: [0->t1, 1->t3], {t3} --> t4".
+func (ht *heapTraversal) getFieldReferences(ref Reference, result ReferenceSet) {
+	ht.visited[ref] = true
+	p := ht.partitions
+	for _, m := range p.PartitionMembers(ref) {
+		if ht.isWithinFunctions(m) {
+			result[m] = true
+		}
+	}
+	rep := p.Representative(ref)
+	for _, r := range p.PartitionFieldMap(rep) {
+		if _, ok := ht.visited[r]; !ok {
+			ht.getFieldReferences(r, result)
+		}
+	}
+}
+
+// Checks whether a taint sink can be reached by a taint source by examining the alias information.
+func IsEARTainted(sink ssa.Instruction, srcRefs ReferenceSet, state *Partitions, callees map[*ssa.Function]bool) bool {
```

Comment:
Is there an expected or required relationship between any of these arguments?

Reply:
These arguments are the required ones used by this public API. We may refactor it to be a public method of heapTraversal:

`func (ht *heapTraversal) IsEARTainted(sink ssa.Instruction, srcRefs ReferenceSet) { ... }`

This will require exposing the `heapTraversal`. Either way works for me.

Reply:
The question was focusing on something slightly different.  We have 3 inter related function cals:
```
callees := earpointer.GetCalleeFunctions(fn, conf.EARCallDepth())
earpointer.GetSrcRefs(s, isTaintField, state, callees)
earpointer.IsEARTainted(sink, srcRefs[src], state, callees)
```
These seem to all be related. Does this function support any srcRefs, state or callees or is there an expected relationship that these are setup exactly like this?

Reply:
Yes `GetSrcRefs()` and `IsEARTainted()` can be combined, while `GetCalleeFunctions()` is also used in `runEAR()`. I actually plan to rewrite the main logic of connecting the sources and sinks to take into account the two examples I gave above. Roughly, given a set of source locations, and a set of sink locations, return the reachable source-sink pairs, possibly limited by the distance between the source and sink. This way we need only one API.
```
  func Reach(sources []Source, sinks []ssa.Instruction) map[Source]map[ssa.Instruction]
```
But this will be done in a future PR, which needs some new unit tests.
Regarding the current limit, we can simply set it to a large number to actually void its effect.

Reply:
I am trying to ask whether these *must* be combined for the function to succeed.

Can I have sequences where these values are unrelated?
```
srcRefs := earpointer.GetSrcRefs(s, isTaintField, state0, callees0)
earpointer.IsEARTainted(sink, srcRefs, state0, callees1)
```
Is this meaningful? Will it crash? Have we baked in a precondition that state0 and state1 must be the same? callees0 and callees1?

Reply:
It won't crash if we use different `state`s or `callees`s, but the results may be not we expect.
The current expectation is: `state`s should be the same since the same heap is traversed for the sources and sinks. In contrast, `callees0` and `callees1` can be different: `callees0` controls the set of sources we are interested in, and `callees1` controls the set of sinks. In fact, using different `callees0` and `callees1` give different heuristics. Basically, `callees0` and `callees1` specify the regions in the heap where we search for the source or sink references.

Reply:
It sounds like state needs additional documentation about what the expectation is then. It sounds like callees do not.